### ISSUE TITLE
[v3] Deserialization of response content, content model type as generic arg and more

### ISSUE THREAD

**Issue by anuchandy** [Azure/autorest-clientruntime-for-java#554]
#### [Details from @JonathanGiles]:
1. Can you explain the difference between RestResponse and HttpResponse? Are they both necessary?
2. It would be useful to learn more about deserialization of response body content. HttpResponse does offer the deserialized body, but it is an Object. It would be nicer if this could be a generic type, if this is expected to be returned to end users. We could require users to provide a function to convert, and we should ideally do it lazily.
3. I’m unclear in the block starting on line 216 of HttpResponseDecoder – nothing in this block is setting a result, so the returned result will be null. I would have presumed the result would be set to wireResponseMap – perhaps a unit test is missing here too?

**Comment by anuchandy** [comment]
I would start with answering 2 which will help to discuss 1 later. 

We redesigned decoding part in last few days. Here is the details -


1. The runtime HttpResponse is low level and should look very close to response type defined by HTTP protocol. It should hold just headers and byte stream representing content. The decoder, decoded body, decoded header should not be part of this type.
2. There should be a “decoding abstraction” for transforming the low level HttpResponse to a Decoded Response. The Decoded Response has one publisher that emits decoded body and another publisher that emits decoded header. Decode upon subscription enables lazy decoding.
3. The decoder (Jackson, Gson or anything else) used by “decoding abstraction” should be pluggable. Runtime’s default decoder implementation will be based on Jackson.
4. The “decoding abstraction” should take type information and additional metadata needed to decode the content and header. This includes:

    * The type of content when response is treated as successful.
[Required] This is the type that the content should be deserialized to when response status code indicates success.
    * The type of content when response is treated as failure.
[Optional] This allows specifying the type that content should be deserialized to when response status code indicates failure. If B  is absent then a response with error code and a body will be deserialized to Object.
    * The expected status codes from service those are treated as success.
[Optional] This allows limiting the status codes to be treated as success. For example -  if as per REST API contract, it returns only 200 for success and if it want client to treat all other status code as failures then this option can be used.
The type of the header.
    * [Optional] This allows extracting a subset of ‘matching’ headers and storing it as properties of an entity instance. Matching headers are the those with name matches with entity property names and/or those with HeaderCollection annotation applied.
 
There are more metadata we can define, such as status code to type mapping. So far we have use-case for only above 4 metadata, so current design defines only those.
 
The design ensures single responsibility for runtime types,  some of the properties of the design includes:
 
1. The Decoded Response won’t emit error if the response status code is error. The body of such response is still needed to be decoded and emits like content of a successful response.
 
2. An SDK/module that deals with raw byte stream and has no concept of decoding can work with the raw HttpResponse, Pipeline and HttpClient.
 
3. Based on deserialization need, the “Decoding abstraction” can be used by Runtime Rest proxy (that process method annotations, prepare request and process response).
 
4. If a SDK/module doesn’t want the overhead of rest proxy but requires decoding can use our “decoding abstraction” by applying a map operation on raw HttpResponse.
 
[PR](https://github.com/Azure/autorest-clientruntime-for-java/pull/560), [commit](
https://github.com/Azure/autorest-clientruntime-for-java/pull/560/commits/d7c0c84eb081aa985fa9cea0a1f4c6748911ba15).

**Comment by anuchandy** [comment]
Answering first question: 
`1. Can you explain the difference between RestResponse and HttpResponse? Are they both necessary?`

In previous comment we discussed the advantage of keeping `HttpResponse` type simple & very close to response defined by Http protocol  i.e. a simple type with headers and byte stream content.

`RestResponse<HeaderT, ContentT>` is more user-friendly type which is returned from methods in proxy interface representing a REST service.

If a REST API returns simple byte stream (e.g. download an image) then return type looks like:

`RestResponse<Headers, Flux<ByteBuf>>`

If a REST API returns an entity say `VirtualMachine` then return type looks like:

`RestResponse<Headers, VirtualMachine>`

If a REST API does not return any entity say a DELELE action, then return type looks like:

`RestResponse<Headers, Void>`


You can see that in case 1 no decoding just stream, in case 2 decoding is required, case 3 no content at all.

**Comment by anuchandy** [comment]
Answering third question

` I’m unclear in the block starting on line 216 of HttpResponseDecoder – nothing in this block is setting a result, so the returned result will be null. I would have presumed the result would be set to wireResponseMap – perhaps a unit test is missing here too?`

This block actually works as expected but it's not readable. Code is not setting result explicitly but it get set since we do action on [`wireResponse`](https://github.com/Azure/autorest-clientruntime-for-java/blob/555ae1547f1b4987470e9e1d697121c8398878c8/client-runtime/src/main/java/com/microsoft/rest/v3/serializer/HttpResponseBodyDecoder.java#L262) variable which is [`result`](https://github.com/Azure/autorest-clientruntime-for-java/blob/555ae1547f1b4987470e9e1d697121c8398878c8/client-runtime/src/main/java/com/microsoft/rest/v3/serializer/HttpResponseBodyDecoder.java#L230).

I've updated the code to make it more explicit.

**Comment by JonathanGiles** [comment]
Thanks for clarifying. I will assume that in our SDK APIs we will therefore use `RestResponse` rather than `HttpResponse` in our public API. The concern I have right now is the verbosity - this is what our azConfig SDK currently defines for a simple getter method: 

`public Mono<RestResponse<Map<String, String>, KeyValue>> getKeyValue(String key, KeyValueFilter filter)`

This is really hard to parse for an end user. It is mostly because of the `Map<String, String>`. I wonder if you have considered hard-coding the type of the headers, rather than requiring the type information as part of the RestResponse generic type? It is also possible that the azConfig code is based on an out-dated version of your runtime?

**Comment by anuchandy** [comment]
Discussed with @JonathanGiles. We will have following two types

```
RestResponse<T>
RestResponseBase<H,T>
```

with `RestResponse<T>` extending from `RestResponseBase<H,T>`.

**Comment by anuchandy** [comment]
@conniey JFYI.

**Comment by JonathanGiles** [comment]
This issue can be closed now I believe

**Comment by anuchandy** [comment]
Closing.
